File: app/src/main/java/org/kinecosystem/kinit/view/adapter/OfferListAdapter.java
package org.kinecosystem.kinit.view.adapter;

import android.content.Context;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;
import java.util.List;
import org.kinecosystem.kinit.R;
import org.kinecosystem.kinit.model.spend.Offer;
import org.kinecosystem.kinit.model.spend.OfferKt;
import org.kinecosystem.kinit.util.ImageUtils;
import org.kinecosystem.kinit.viewmodel.spend.SpendViewModel;

public class OfferListAdapter extends RecyclerView.Adapter<OfferListAdapter.ViewHolder> {


    private List<Offer> offerList;
    private Context context;
    private SpendViewModel model;

    public OfferListAdapter(Context context, SpendViewModel model) {
        this.context = context;
        this.model = model;
        offerList = model.getOffers();
    }

    @Override
    public OfferListAdapter.ViewHolder onCreateViewHolder(ViewGroup parent,
        int viewType) {
        View view = LayoutInflater.from(parent.getContext())
            .inflate(R.layout.offer_card, parent, false);
        ViewHolder vh = new ViewHolder(context, view);
        return vh;
    }

    @Override
    public void onBindViewHolder(ViewHolder holder, int position) {
        holder.bind(offerList.get(position));
        holder.view.setOnClickListener(view -> model.onItemClicked(position));
    }

    @Override
    public int getItemCount() {
        return offerList.size();
    }


    public static class ViewHolder extends RecyclerView.ViewHolder {

        private TextView title, providerName, reward;
        private ImageView imageView, logoView;
        private Context context;
        public View view;

        public ViewHolder(Context context, View view) {
            super(view);
            this.view = view;
            this.context = context;
            title = view.findViewById(R.id.offer_title);
            providerName = view.findViewById(R.id.provider_name);
            reward = view.findViewById(R.id.offer_reward);
            imageView = view.findViewById(R.id.offer_image);
            logoView = view.findViewById(R.id.offer_logo);
        }

        public void bind(Offer offer) {
            title.setText(offer.getTitle());
            providerName.setText(offer.getProvider().getName());
            if (OfferKt.isP2p(offer)) {
                reward.setVisibility(View.INVISIBLE);
            } else {
                reward.setText(offer.getPrice() + " KIN");
            }
            ImageUtils.loadImageIntoView(context, offer.getImageUrl(), imageView);
            ImageUtils.loadImageIntoView(context, offer.getProvider().getImageUrl(), logoView);
        }
    }
}

File: app/src/main/java/org/kinecosystem/kinit/view/adapter/OfferListAdapter.kt
package org.kinecosystem.kinit.view.adapter

import android.content.Context
import android.databinding.Observable
import android.support.v4.content.ContextCompat
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import org.kinecosystem.kinit.R
import org.kinecosystem.kinit.model.spend.Offer
import org.kinecosystem.kinit.model.spend.isAvailable
import org.kinecosystem.kinit.model.spend.isP2p
import org.kinecosystem.kinit.util.ImageUtils
import org.kinecosystem.kinit.viewmodel.spend.SpendViewModel

class OfferListAdapter(private val context: Context, private val model: SpendViewModel)
    : RecyclerView.Adapter<OfferListAdapter.ViewHolder>() {


    private var offers: List<Offer> = model.offers()
    private val propertyChangeCallBack = object : Observable.OnPropertyChangedCallback() {
        override fun onPropertyChanged(sender: Observable?, propertyId: Int) {
            refresh()
        }
    }

    fun refresh() {
        offers = model.offers()
        model.onDataLoaded()
        notifyDataSetChanged()
    }

    override fun onCreateViewHolder(parent: ViewGroup,
                                    viewType: Int): OfferListAdapter.ViewHolder {
        val view = LayoutInflater.from(parent.context)
                .inflate(R.layout.offer_card, parent, false)
        return ViewHolder(context, view)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        val offer: Offer = offers[position]
        holder.bind(offer)
        if (offer.isAvailable()) {
            holder.view.setOnClickListener { model.onItemClicked(offer, position) }
        } else {
            holder.view.setOnClickListener { null }
        }
    }

    override fun getItemCount(): Int {
        return offers.size
    }

    override fun onDetachedFromRecyclerView(recyclerView: RecyclerView) {
        super.onDetachedFromRecyclerView(recyclerView)
        model.offersRepository.offers.removeOnPropertyChangedCallback(propertyChangeCallBack)
    }

    override fun onAttachedToRecyclerView(recyclerView: RecyclerView) {
        super.onAttachedToRecyclerView(recyclerView)
        model.offersRepository.offers.addOnPropertyChangedCallback(propertyChangeCallBack)
    }


    class ViewHolder(private val context: Context, val view: View) : RecyclerView.ViewHolder(view) {
        private val title: TextView = view.findViewById(R.id.offer_title)
        private val providerName: TextView = view.findViewById(R.id.provider_name)
        private val reward: TextView = view.findViewById(R.id.offer_reward)
        private val imageView: ImageView = view.findViewById(R.id.offer_image)
        private val logoView: ImageView = view.findViewById(R.id.offer_logo)
        private val unavailableTitle: TextView = view.findViewById(R.id.unavailable_title)

        fun bind(offer: Offer) {
            title.text = offer.title
            providerName.text = offer.provider!!.name
            if (offer.isP2p()) {
                reward.visibility = View.INVISIBLE
            } else {
                reward.text = offer.price!!.toString() + " KIN"
            }
            if (offer.isAvailable()) {
                reward.setTextColor(ContextCompat.getColor(context, R.color.blue))
            } else {
                reward.setTextColor(ContextCompat.getColor(context, R.color.grey_dark_medium))
            }
            ImageUtils.loadImageIntoView(context, offer.imageUrl, imageView)
            ImageUtils.loadImageIntoView(context, offer.provider.imageUrl, logoView)
            offer.unavailableReason?.let {
                unavailableTitle.text = it
                imageView.alpha = 0.2f
            } ?: run {
                unavailableTitle.text = ""
                imageView.alpha = 1f
            }
        }
    }
}